test/src/test/integration/link_git_protocol.rs

```diff
@@ -6,91 +6,106 @@
 use std::{
     collections::BTreeSet,
     io,
-    path::Path,
+    path::{Path, PathBuf},
     sync::{atomic::AtomicBool, Arc},
 };
 
 use bstr::ByteSlice as _;
 use futures::{AsyncReadExt as _, TryFutureExt as _};
+use git::{
+    prelude::*,
+    refs::transaction::{Change, PreviousValue, RefEdit},
+};
+use git_repository as git;
 use link_git_protocol::{fetch, ls, packwriter, upload_pack, ObjectId, PackWriter, Ref};
 use tempfile::{tempdir, TempDir};
 
 fn upstream() -> TempDir {
     let tmp = tempdir().unwrap();
 
-    let repo = git2::Repository::init_bare(&tmp).unwrap();
-    let auth = git2::Signature::now("apollo", "apollo@cree.de").unwrap();
-
-    let tree = {
-        let empty = repo.treebuilder(None).unwrap();
-        let oid = empty.write().unwrap();
-        repo.find_tree(oid).unwrap()
-    };
-    let base = {
-        let oid = repo
-            .commit(
-                Some("refs/namespaces/foo/refs/heads/main"),
-                &auth,
-                &auth,
-                "initial",
-                &tree,
-                &[],
-            )
-            .unwrap();
-        repo.find_commit(oid).unwrap()
+    let repo = git::init_bare(&tmp).unwrap().into_easy();
+    let auth = git::actor::Signature::now_local_or_utc("apollo", "apollo@cree.de");
+
+    let empty_tree_id = repo
+        .write_object(&git::objs::Tree::empty().into())
+        .unwrap()
+        .detach();
+    let base_id = {
+        repo.commit(
+            "refs/namespaces/foo/refs/heads/main",
+            "initial",
+            auth.clone(),
+            auth.clone(),
+            empty_tree_id,
+            git::commit::NO_PARENT_IDS,
+        )
+        .unwrap()
     };
-    let next = repo
+    let next_commit_id = repo
         .commit(
-            Some("refs/namespaces/foo/refs/heads/next"),
-            &auth,
-            &auth,
+            "refs/namespaces/foo/refs/heads/next",
             "ng",
-            &tree,
-            &[&base],
+            auth.clone(),
+            auth,
+            empty_tree_id,
+            [base_id],
         )
-        .unwrap();
+        .unwrap()
+        .detach();
     repo.reference(
```

Comment:
I always found it a bit odd that `reference` _creates_ a reference, but well names

Reply:
Me too! And it was aptly called `create_reference` before. But as I worked with this it dawned on me how `libgit2` (or `git2`) is thinking and I started liking it and thought it was beneficial to not unnecessarily change names.

This also means that `reference**s**` now creates an intermediate object for iterators, and you guessed it, that was called `iter_references()` before I took the U-turn there as well.

I don't know if it's _good_ to be so `git2` inspired there, and already thought providing an alternative name for the same functionality potentially even behind feature 

[2000 words later]… names _are_ hard.